zkevm-circuits/src/state_circuit/constraint_builder.rs

```diff
@@ -226,10 +231,80 @@ impl<F: Field> ConstraintBuilder<F> {
         // TODO: Missing constraints
     }
 
+    fn build_tx_log_constraints(&mut self, q: &Queries<F>) {
+        self.require_equal(
+            "is_write is always true for TxLog",
+            q.is_write.clone(),
+            1.expr(),
+        );
+        self.require_in_set(
+            "field_tag in TxLogFieldTag range",
+            q.field_tag(),
+            set::<F, TxLogFieldTag>(),
+        );
+        // reset log_id when tx_id increases
+        let no_tag_change = 1.expr() - q.tag_change();
+        self.condition(q.id_change() * no_tag_change, |cb| {
+            cb.require_zero(
+                "reset log_id to one when tx_id increases",
+                q.tx_log_id(false) - 1.expr(),
+            );
+            // constrain first field_tag is Address when tx id increases
+            cb.require_equal(
+                "first field_tag is Address",
+                q.field_tag_matches(TxLogFieldTag::Address),
+                1.expr(),
+            );
+            // tx_id increases by 1
+            cb.require_equal("tx_id increases by 1", q.id_change(), 1.expr());
+        });
+
+        self.condition(q.field_tag_matches(TxLogFieldTag::Address), |cb| {
+            cb.require_zero("index is zero for address ", q.tx_log_index(false))
+        });
+
+        // increase log_id when tag changes to Address if tx id stay same
+        self.condition(
+            q.is_id_unchanged.clone()
+                * q.tag_matches(RwTableTag::TxLog, true)
+                * q.field_tag_matches(TxLogFieldTag::Address),
+            |cb| cb.require_equal("log_id = pre_log_id + 1", q.tx_log_id(false), 1.expr()),
+        );
+
+        // if tag Topic appear, topic_index in range [0,4),can only increase by one when
+        // tag stays same.
+        self.condition(q.field_tag_matches(TxLogFieldTag::Topic), |cb| {
+            let topic_index = q.tx_log_index(false);
+            cb.require_zero(
+                "topic_index in range [0,4) ",
+                topic_index.clone()
+                    * (1.expr() - topic_index.clone())
+                    * (2.expr() - topic_index.clone())
+                    * (3.expr() - topic_index),
+            )
+        });
+        // constrain index is increasing by 1 when field_tag stay same
+        // let no_field_tag_change= 1.expr() - q.field_tag_change();
+        self.condition(
+            q.tag_matches(RwTableTag::TxLog, true) * q.is_field_tag_unchanged.clone(),
+            |cb| {
+                cb.require_equal(
+                    "index = pre_index + 1",
+                    q.tx_log_index(false),
+                    q.tx_log_index(true) + 1.expr(),
+                )
+            },
+        );
+    }
```

Comment:
is there any constraint to check "in TxLog, with same txid,  log_id will not change if field_tag != Address"?

Reply:
thanks , I think it makes sense, but this is not in origin log state constrain  spec, I will try to add it into spec first :), then handle the circuit side .

Reply:
added in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/552/commits/06c2826936e0b36f460550d7d8f1f1856c779e7a thanks !!